webrtc/RTCPeerConnection-ondatachannel.html

```diff
@@ -5,180 +5,388 @@
-      attribute set to channel, must be created and dispatched at the given
-      target.
-
-      interface RTCDataChannelEvent : Event {
-        readonly attribute RTCDataChannel channel;
-      };
-   */
-  async_test(t => {
-    const localPc = new RTCPeerConnection();
-    t.add_cleanup(() => localPc.close());
-    const remotePc = new RTCPeerConnection();
-
-    t.add_cleanup(() => remotePc.close());
-
-    let eventCount = 0;
-
-    const onDataChannel = t.step_func_done(event => {
-      eventCount++;
-      assert_equals(eventCount, 1,
-        'Expect data channel event to fire exactly once');
-
-      assert_true(event instanceof RTCDataChannelEvent,
-        'Expect event to be instance of RTCDataChannelEvent');
-
-      assert_equals(event.bubbles, false);
-      assert_equals(event.cancelable, false);
-
-      const { channel } = event;
-      assert_true(channel instanceof RTCDataChannel,
-        'Expect channel to be instance of RTCDataChannel');
-
-      const { readyState } = channel;
-
-      // The spec requires readyState to be connecting at first,
-      // but it may quickly change to open before the callback
-      // is invoked, especially with local connections.
-      assert_true(readyState === 'connecting' || readyState === 'open',
-        'Expect channel ready state to be either connecting or open');
-    });
+'use strict';
+
+// Test is based on the following revision:
+// https://rawgit.com/w3c/webrtc-pc/1cc5bfc3ff18741033d804c4a71f7891242fb5b3/webrtc.html
+
+// The following helper functions are called from RTCPeerConnection-helper.js:
+// exchangeIceCandidates
+// doSignalingHandshake
+// createDataChannelPair
+
+/*
+  6.2.  RTCDataChannel
+    When an underlying data transport is to be announced (the other peer created a channel with
+    negotiated unset or set to false), the user agent of the peer that did not initiate the
+    creation process MUST queue a task to run the following steps:
+      2. Let channel be a newly created RTCDataChannel object.
+      7. Set channel's [[ReadyState]] to open (but do not fire the open event, yet).
+      8. Fire a datachannel event named datachannel with channel at the RTCPeerConnection object.
+
+  6.3.  RTCDataChannelEvent
+    Firing a datachannel event named e with an RTCDataChannel channel means that an event with the
+    name e, which does not bubble (except where otherwise stated) and is not cancelable (except
+    where otherwise stated), and which uses the RTCDataChannelEvent interface with the channel
+    attribute set to channel, MUST be created and dispatched at the given target.
+
+    interface RTCDataChannelEvent : Event {
+      readonly attribute RTCDataChannel channel;
+    };
+ */
+promise_test(async (t) => {
+  const resolver = new Resolver();
+  const pc1 = new RTCPeerConnection();
+  const pc2 = new RTCPeerConnection();
+  t.add_cleanup(() => pc1.close());
+  t.add_cleanup(() => pc2.close());
```

Comment:
Note that if this promise_test was in its own file, we could use a single file test, which has some benefits:
- no need for promise_test, step_func wrappers, resolver the test just needs to call 'done()' at some point.
- no need for t.add_cleanup(() => pc1.close()); since the browser would navigate for the next test.

Reply:
I don't think that's a good idea since the file may be extended with further tests at some point.